cadc-data-ops-fits/src/main/java/org/opencadc/fits/slice/NDimensionalSlicer.java

```diff
@@ -0,0 +1,587 @@
+            // requested value.  Boxing extVer into a new Integer() alleviates a NullPointerException from possibly
+            // occurring.
+            return extName.equalsIgnoreCase(extensionName)
+                   && (((extensionVersion == null) && (extVer == 1))
+                       || (Integer.valueOf(extVer).equals(extensionVersion)));
+        }
+
+        return false;
+    }
+
+    /**
+     * Obtain the overlapping indexes of matching HDUs.  This will return a unique list of Integers in file order,
+     * rather than the order that they were requested.  Since
+     *
+     * @param fits                  The Fits object to scan.
+     * @param extensionSliceValues  The requested values.
+     * @return  An Map of overlapping hduIndex->slice[], or empty Map.  Never null.
+     * @throws FitsException if the header could not be read
+     * @throws IOException   if the underlying buffer threw an error
+     */
+    private Map<Integer, Slices.ExtensionSliceValue[]> getOverlap(final Fits fits,
+                                                                  final Slices.ExtensionSliceValue[] extensionSliceValues)
+            throws FitsException, IOException {
+
+        // A Set is used to eliminate duplicates from the inner loop below.
+        final Map<Integer, Slices.ExtensionSliceValue[]> overlapHDUIndexesSlices = new LinkedHashMap<>();
+
+        // Walk through the cache first.
+        final int hduCount = fits.getNumberOfHDUs();
+        int matchCount = 0;
+        int hduIndex;
+        for (hduIndex = 0; hduIndex < hduCount; hduIndex++) {
+            final BasicHDU<?> hdu = fits.getHDU(hduIndex);
+            final Slices.ExtensionSliceValue[] overlapSlices = getOverlap(hdu, hduIndex, extensionSliceValues);
+            if (overlapSlices.length > 0) {
+                overlapHDUIndexesSlices.put(hduIndex, overlapSlices);
+                matchCount += overlapSlices.length;
+            }
+        }
+
+        // Read in each HDU that is not in cache.  This will only read until it doesn't need to anymore.
+        BasicHDU<?> hdu;
+        while ((matchCount < extensionSliceValues.length) && ((hdu = fits.readHDU()) != null)) {
+            final Slices.ExtensionSliceValue[] overlapSlices = getOverlap(hdu, hduIndex, extensionSliceValues);
+            if (overlapSlices.length > 0) {
+                overlapHDUIndexesSlices.put(hduIndex, overlapSlices);
+                matchCount += overlapSlices.length;
+            }
+            hduIndex++;
+        }
+
+        return overlapHDUIndexesSlices;
+    }
+
+    private Slices.ExtensionSliceValue[] getOverlap(final BasicHDU<?> hdu, final int hduIndex,
+                                                    final Slices.ExtensionSliceValue[] extensionSliceValues)
+            throws FitsException {
+        final List<Slices.ExtensionSliceValue> overlappingSlices = new ArrayList<>();
+
+        if (hdu != null) {
+            if (!(hdu instanceof ImageHDU)) {
```

Comment:
I wasn'ty going to review "correct output" because round trip tests, fitsverify, and eventually astronomers are better at that than code review can be, but since you moved this `if (ImageHDU)` in here I will comment. I agree that it has to be in here because errors have to be thrown and reported before opening output stream so that means in here.

This method **should not blindly** throw away no-image extensions. If the caller is only extracting an extension then all extension types can be output -- whether or not an extension can be returned depends on the applicability of the operation to the extension. That will be a combination of content and whatever code we implement. 

For now:
* extension extraction should work for all extensions, extension not found? **TBD but for now IllegalArgumentException**
* pixel cutouts should work for image extensions, **else: UnsupportedOperationException**

In near future:
* sky cutouts work for image extensions with complete-enough WCS, **else: skip**

In future:
* some sort of row,column (~pixel) cutouts could be implemented to work on FITS tables
* sky cutout could be implemented on FITS tables if columns contain coordinates (eg event lists)